Sources/NIOCore/ByteBuffer-core.swift

```diff
@@ -825,6 +825,23 @@ public struct ByteBuffer {
         }
         return true
     }
+    
+    /// Shrinks the buffers capacity to a `desiredCapacity` without altering readably bytes.
+    ///
+    /// - Parameter desiredCapacity: The desired capacity for the buffers capacity to be shrunken to
+    /// - Returns: Bool indicating whether the buffer capacity has been shrunk to the desiredCapacity.
+    @inlinable
+    @discardableResult public mutating func shrinkBufferCapacity(to desiredCapacity: Int) -> Bool {
+        guard desiredCapacity < capacity, desiredCapacity > readableBytes else {
```

Comment:
We should compare to the effective requested capacity, which will be `capacity == 0 ? 0 : capacity.nextPowerOf2ClampedToMax()`.

Now is also probably a good time to factor that computation out to a `static func` on `ByteBuffer`, as we're probably going to want to tweak it (see #2729).

Reply:
Sorry, I don't understand this. 

If desired capacity is 1023 and the current capacity is also 512, would comparing with capacity.nextPowerOf2 not mean that the condition passes and this operation actually reallocates the buffers capacity by expanding it? 
This is assuming the below implementation method which is what I got from your comment. 
`guard desiredCapacity < capacity. nextPowerOf2(), desiredCapacity > readableBytes`

Also, since Capacity it an integer, I think it can only call nextPowerOf2 and not nextPowerOf2ClampedToMax. Meaning we do not need the `capacity == 0 ? 0` as it is already done in `nextPowerOf2` by a 
`
guard self != 0 else {
            return 1
        }
`

Reply:
The check is to move `desiredCapacity` to `nextPowerOf2`, not `currentCapacity`. In that case, the desired capacity would be 1024, which is larger than current capacity at 512.

Reply:
Hey @Lukasa. I think I have addressed this in the latest commit. Please take a look when you find some time.

Reply:
Bump